src/normalization.md

```diff
@@ -0,0 +1,314 @@
+
+Here is a rough overview of the different entry points to normalization in the compiler:
+- `infcx.at.(deeply_/structurally_)?normalize`
+- `tcx.normalize_erasing_regions`
+- `infcx.query_normalize`
+- `traits::normalize_with_depth(_to)`
+
+#### `InferCtxt` based normalization and derivatives
+
+The [`InferCtxt`][infcx] type exposes the "main" ways to normalize during analysis: [`normalize`][normalize], [`deeply_normalize`][deeply_normalize] and [`structurally_normalize`][structurally_normalize]. These functions are often wrapped and re-exposed on various `InferCtxt` wrapper types, such as [`FnCtxt`][fcx] or [`ObligationCtxt`][ocx] with minor API tweaks to handle some arguments or parts of the return type automatically.
+
+These methods of normalization are intended to be used in contexts where inference variables are expected to be encountered and also are able to be used in situations where regions are important to consider. 
+
+In the old solver: 
+- `infcx.at.normalize`/`structurally_normalize` will fold over the provided value normalizing all aliases inside of it. 
+- `deeply_normalize`: Folds over the provided value normalizing all aliases, erroring on ambiguous aliases[^1].
+
+There is not much of a distinction between the three methods of normalizing with `InferCtxt`. The primary reason for there to exist all three is as a transitionary step between the old and new solver, because in the new solver there is a distinction between deeply normalizing and structurally normalizing, and the old solver behaviour no longer exists. 
+
+In the new solver:
+- `infcx.at.normalize`: A no-op, `deeply_normalize` or `structurally_normalize` should be used instead. This method only exists for the old solver.
+- `deeply_normalize`: Folds over the provided value normalizing all aliases, erroring on ambiguous aliases[^1].
+- `structurally_normalize`: Normalizes an alias to a rigid type or an inference variable, does nothing if the type is not an alias.
+
+When the new solver is stabilized the `infcx.at.normalize` function will be removed and everything will have been migrated to either deeply normalize or structurally normalize. In cases where normalization is required in both old and new solvers, the `structurally_normalize` function is often the right choice. Otherwise, if normalization is only needed in the old solver then `infcx.at.normalize` is often the right choice.
+
+`structurally_normalize` should be thought of as only normalizing the "outer layer" of a type to a rigid type or inference variable. Any types inside of the normalized form may still be unnormalized (even if an alias *was* actually normalized[^2]). `structurally_normalize` is named such as it is typically called before looking at the "structure" of a type (e.g. by matching on `TyKind`/`ConstKind`).
```

Comment:
i think you probably should move this up without yet talking about impl details.

shallow vs deep normalization is an important distinction